### PR TITLE
Fixed variable type changing due to serialization

### PR DESCRIPTION
**Pull request by clarkmcc**

### CODE REVIEW

File: interpreter/src/ser.rs [hunk 1]
```diff
@@ -561,31 +561,6 @@ impl ser::Serializer for KeySerializer {
     }
 }
 
-impl Serialize for Value {
```
Comment (alexsnaps) [inline interpreter/src/ser.rs:564]: A little surprised this was there… What was the intended use-case for serializing `Value` instances? 
Might be the thing you actually wanted to ponder, i.e. the repercussions of removing this ?

File: interpreter/src/objects.rs [hunk 2]
```diff
@@ -151,6 +142,13 @@ impl<T: Serialize> TryIntoValue for T {
     }
 }
 
+impl TryIntoValue for Value {
```
Comment (alexsnaps) [inline interpreter/src/objects.rs:145]: Given `Context::add_variable`'s signature I think this makes sense. I probably should have looked at other methods on context or read the code, but it looked the natural fit for the job… Tho tbh, I got thrown off by the `Result` returned… but happily ignored it for now and moved on. 

Not a real proposal, as I may real be off wrt how people use this, but wondering if `::add_variable_from_value` shouldn't be `::add_variable` and have something along the `::try_add_variable` for fallible `TryIntoValue` values… 
On a more generic note, I wonder if folding the `.try_into()` within the method might be blurring the lines here. As it can fail, the user it left with dealing with the `Err` case, which is the _only_ error case possible. Which now splits the contract to some extent. So wouldn't it be clearer to leave it to the user to call `try_into_value()`? Which would also reduce the code footprint of the resulting binary (tho the user could take that upon themselves if they cared)… Anyways, wondering 🤔

File: interpreter/src/ser.rs (hunk 1 above)
Comment (clarkmcc) [reply interpreter/src/ser.rs:564]: I'm pretty sure it was a poorly thought out solution for allowing the add_variable method to support both serializable types and values -- by making values serializable. The more I think about it, the more I'm comfortable with my changes here, I think they accomplish the same objective.

Comment (alexsnaps) [reply interpreter/src/ser.rs:564]: Cool! Live and learn (and rinse & repeat! 😉)

File: interpreter/src/objects.rs (hunk 2 above)
Comment (alexsnaps) [reply interpreter/src/objects.rs:145]: Another question, and I haven't tried to answer it myself, sorry... Why `trait TryIntoValue` and not core's `Try(From|Into)`?

Comment (clarkmcc) [reply interpreter/src/objects.rs:145]: > Not a real proposal, as I may real be off wrt how people use this, but wondering if ::add_variable_from_value shouldn't be ::add_variable and have something along the ::try_add_variable for fallible TryIntoValue values…

I'd be comfortable with this. I think the original intent was to not make a breaking API change (except to introduce the result). Making `add_variable` accept really anything wouldn't require any code migrations from users. That being said, as a pre-1.0 project, I think idiom and expectation is probably more important than backwards compatibility.

> On a more generic note, I wonder if folding the .try_into() within the method might be blurring the lines here.

[This](https://github.com/clarkmcc/cel-rust/pull/33#issuecomment-1946497968) was a point of discussion in the initial PR. My argument at the time was that doing implicit serialization is consistent with the [principle of least surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment), in other words, if the user was going to have to have their value serialized at some point, and if that serialization error needs to be handled, why not do as much as possible for the user so they don't need to know to do it on their own. I cited these examples[1] of cases in the wild where `T: Serialize` is accepted instead of the already-serialized result.

But clearly both you and the original PR author at least initially expected a different API, so I'd be happy to hear your counter-arguments for this rationale.

> Another question, and I haven't tried to answer it myself, sorry... Why trait TryIntoValue and not core's Try(From|Into)?

For one, we would need to do something like this `impl<T: Serialize> TryFrom<T> for Value` and that is not allowed because it conflicts with the blanket implementation `impl<T, U> TryFrom<U> for T where U: std::convert::Into<T>;`
- The blanket implementation covers any type `T` and any type `U` where `U` can be converted into `T`.
- The proposed `TryFrom` implementation also covers any type `T` that implements Serialize.
- Since `Serialize` is a widely implemented trait, many types will satisfy both conditions: they can be converted into other types (via Into) and they implement Serialize.

[1]: https://sourcegraph.com/search?q=context:global+/T:+Serialize%3E%5C(/&patternType=keyword&sm=0

Comment (alexsnaps) [reply interpreter/src/objects.rs:145]: >I cited these examples[1](https://sourcegraph.com/search?q=context:global+/T:+Serialize%3E%5C(/&patternType=keyword&sm=0) of cases in the wild where T: Serialize is accepted instead of the already-serialized result.

Quickly looking through some of these tho, I found 2 patterns I think: either a utility functions, which aims at encapsulating the serialization logic; or method, that depending on the `self`'s state, can avoid the serialization altogether in some cases. But I'm sure there are other cases.

I think this here differs in the sense that the fallible code path is always invoked (or at least today is), and the primary use of the method is to mutate the context. Again, I don't have a huge preference here. I probably would have done it differently, i.e. leave both the handing the serialization and dealing with the error to the user, more so that `add_variable`, rightly so, takes ownership of the `value` and the error case leaving the user with a `Box<dyn Error>` if I read this right. 

>as a pre-1.0 project, I think idiom and expectation is probably more important than backwards compatibility.

I strongly agree with that statement tho... So at this stage I think it might be better to "flag" this as something worth investigating a little more, as to what the best, yet flexible API for a user would be. And I know I don't have enough experience with CEL and its usages to be able to form a half valid opinion at this stage. But willing to keep on "playing" with the API to possibly come up with a proposal and matching usages to showcase pros & cons of any given approach... 

So tl;dr I wouldn't necessarily change anything beyond what you have done at this stage.

Comment (clarkmcc) [reply interpreter/src/objects.rs:145]: Understood. Yeah regardless, I think this feedback should be addressed as a separate PR. Feel free to open an issue when you're ready and we can continue the discussion on a better API design. I wasn't really thrilled with the proposed alternative at the time
https://github.com/clarkmcc/cel-rust/blob/2cb75cb3cc985583c199ea4891ddd5450d6b1beb/example/src/serde.rs#L20

Maybe it's the right solution, maybe we instead implement a `SerializeToValue` trait to attach methods to any `T: Serialize` and perhaps that's cleaner. Not sure.

Comment (alexsnaps) [reply interpreter/src/objects.rs:145]: Yeah... `SerializeToValue` is an interesting idea. There is some talk about protobuf & some JSON too in the cel spec. The former is actually close that what I'm dealing with, which is injecting `Value` in the `Context` that come from [envoy](https://www.envoyproxy.io/), which is the host to my wasm module, integrated using [`proxy-wasm`](https://github.com/proxy-wasm/proxy-wasm-rust-sdk). So my values come as `Vec<u8>`, which could be quite a few things, but also some protobuf message. Anyways don't want to pollute this thread to much, but based on what I've seen, pulling values in some form into the context of a cel expression's evaluation seems to be a common thing. Some forms _seem_ to be better supported/spec'ed than others... e.g. protobuf & json. If that's indeed the case, then I think these would be the first usecase I'd look into and see how they'd play along with the Rust ecosystem around these serialized forms.

File: interpreter/src/ser.rs (hunk 1 above)
Comment (jplock) [reply interpreter/src/ser.rs:564]: Hi, the removal of this broke https://github.com/clarkmcc/cel-rust/pull/40.  What I've been doing is calling:

```rust
let result: serde_json::Value = serde_json::to_value(value).map_err(|err| anyhow!("Unable to serialize JSON value: {}", err))?;
```

Where `value` is a `cel_interpreter::Value`. Is there a better/easier way I can convert a CEL `Value` into a `serde_json::Value`?

Comment (alexsnaps) [reply interpreter/src/ser.rs:564]: So there is one thing I'm wondering here... _I think_ it relates [this issue](https://github.com/clarkmcc/cel-rust/issues/73), ... and I could be wrong as I haven't looked how the golang reference implementation goes about it. But shouldn't these things be considered "together" and have a "symmetric" solution to get `Value`s in & out of "the CEL space" into others? Mostly thinking protobuf and JSON, as these are called out explicitly in the language spec? Feel free to ignore, as I know I don't have a handle on the problem space quite yet, but I know I would like to dig into it (as my use case is CEL & Protobuf).

Comment (clarkmcc) [reply interpreter/src/ser.rs:564]: @alexsnaps I think you're on the right track -- when you want get a strongly typed output, I don't believe the correct solution here is to serialize the CEL output to JSON, then deserialize from JSON to a struct. I think we want some mechanism to turn `Value` into structs and then you can do whatever you want with them.

Looking at cel-go this is what you get back as the result of an expression. You wouldn't serialize this directly to JSON either. You'd switch on type, or you'd convert directly to a Go struct that was registered to the interpreter.
https://github.com/google/cel-go/blob/37d5b961f5da8ca2a21ccb565af63faaf84ed2fa/common/types/ref/reference.go#L37-L63

Comment (clarkmcc) [reply interpreter/src/ser.rs:564]: @jplock if you haven't already, you might pin your dependency to `0.8.0` to work around this change until we can nail down the API for this sort of thing going forward. Thanks for bringing up the use-case!

Comment (clarkmcc) [reply interpreter/src/ser.rs:564]: @jplock would you mind giving [this branch](https://github.com/clarkmcc/cel-rust/pull/77) a try?

Comment (jplock) [reply interpreter/src/ser.rs:564]: I tested it, works great. Thank you!

Comment (fore5fire) [reply interpreter/src/ser.rs:564]: I'll throw out there that I have a use case that's broken by this change as well: I'm using serde to avoid thousands of lines of binding code for conversion of many structs to CEL (I haven't quite wrapped my head around writing custom marcos yet), but there are a few fields that accept arbitrary input (as long as it can be represented in both TOML and CEL). I was using `cel_interpreter::Value` as the type for those fields, but after updating these no longer work with the `Serialize` derive.

For now I'll just create a wrapper type and copy this removed implementation to my project, but fwiw the ability to nest `Value` in a type and use the `Serialize` derive was useful to me, so much so that I wrote a PR to fix almost the exact issue this PR did except when using serde
